feat: add access token updated event

porcellus opened the pull request:
## Summary of change
Add an event that signals if the access token has been updated to support updating the access token payload in the session context of the react lib.

## Related issues
- supertokens/supertokens-auth-react/issues/397

## Test Plan

Added a test case for both Axios and fetch.

## Documentation changes

- [ ] TODO: update `handle-event.mdx` in the session recipe

## Checklist for important updates
- [x] Changelog has been updated
- [x] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
   - Along with the associated array in `lib/ts/version.ts`
- [x] Changes to the version if needed
   - In `package.json`
   - In `package-lock.json`
   - In `lib/ts/version.ts`
- [x] Had run `npm run build-pretty`
- [x] Had installed and ran the pre-commit hook
- [x] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

rishabhpoddar commented on lib/ts/axios.ts:
```diff
@@ -179,6 +181,9 @@ export function responseInterceptor(axiosInstance: any) {
                 let frontToken = response.headers["front-token"];
                 if (frontToken !== undefined) {
                     await FrontToken.setItem(frontToken);
+                    if (!isNewSession) {
+                        onTokenUpdate();
+                    }
```
This makes sense, however there could be another (maybe better) way of doing this: In `FrontToken.setItem`, we check if a token already exists (so a session exists), and then in case the new access token payload is different than what is saved, we fire this event. We would have to be careful about not doing that during session refresh.

What do you think?

rishabhpoddar commented on test/axios2.test.js:
```diff
@@ -1271,4 +1271,102 @@ describe("Axios AuthHttpRequest class tests", function() {
+                "LOGIN_FINISH",
+                "ACCESS_TOKEN_PAYLOAD_UPDATED",
+                "UPDATE1_FINISH",
+                "REFRESH_SESSION",
+                "ACCESS_TOKEN_PAYLOAD_UPDATED",
+                "REFRESH_FINISH",
```
Shouldn't this not have the `ACCESS_TOKEN_PAYLOAD_UPDATED` event?

porcellus replied:
I later renamed this event to `ACCESS_TOKEN_UPDATED` since that is more accurate, and the next call after a refresh does update the access token.

rishabhpoddar replied:
> and the next call after a refresh does update the access token.

I don't understand this. You removed the `ACCESS_TOKEN_PAYLOAD_UPDATED`, therefore the older `ACCESS_TOKEN_PAYLOAD_UPDATED` was incorrect?

rishabhpoddar commented:
@porcellus there are a few conflicts too.

rishabhpoddar commented on CHANGELOG.md:
```diff
@@ -9,6 +9,20 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+### Adds
+-   New FDI support (v1.13)
+-   Workflow to verify if pr title follows conventional commits
+
```
Why is the changelog showing from previous PRs too?

rishabhpoddar replied:
nvm.. ignore this comment

rishabhpoddar commented on lib/ts/axios.ts:
```diff
@@ -361,6 +361,8 @@ export default class AuthHttpRequest {
                         }
                         let frontToken = response.headers["front-token"];
                         if (frontToken !== undefined) {
+                            // this is also done by the response interceptor so
+                            // not firing the access token updated event
```
This comment is not necessary anymore.

rishabhpoddar commented on lib/ts/fetch.ts:
```diff
@@ -558,7 +565,9 @@ export async function setIdRefreshToken(idRefreshToken: string | "remove", statu
         AuthHttpRequest.config.onHandleEvent({
             action: "SESSION_CREATED"
         });
+        return true;
     }
+    return false;
```
We do not need to return a bool from this function anymore.

rishabhpoddar commented on lib/ts/fetch.ts:
```diff
@@ -685,5 +698,13 @@ export async function setFrontToken(frontToken: string | undefined) {
     }
 
+    const oldToken = await getFrontToken();
+    if (oldToken !== null && frontToken !== undefined) {
+        const oldPayload = parseFrontToken(oldToken).up;
+        const newPayload = parseFrontToken(frontToken).up;
+        if (JSON.stringify(oldPayload) !== JSON.stringify(newPayload)) {
+            onTokenUpdate();
+        }
+    }
```
This may get triggered during session refresh as well - if the refresh API causes a change in the access token payload. I think it should fire even during session refresh - while from auth-react's point of view this will cause a double update, semantically, firing the `ACCESS_TOKEN_PAYLOAD_UPDATED ` event still makes sense.

What do you think?

rishabhpoddar replied:
We should also add a test for this where a session refresh causes a change in the access token payload and then on session refresh, we check that `ACCESS_TOKEN_PAYLOAD_UPDATED` is fired.

porcellus replied:
Yep, I think that makes sense. It already works that way, so I've added a test for it.

porcellus commented:
>I don't understand this. You removed the ACCESS_TOKEN_PAYLOAD_UPDATED, therefore the older ACCESS_TOKEN_PAYLOAD_UPDATED was incorrect?

So: I've renamed the event to `ACCESS_TOKEN_UPDATED` which is what I was checking in the test, and that was correct - and later I've removed it, since it doesn't make sense if we are only firing the event if the payload was updated (and I've reverted the rename)

rishabhpoddar commented on test/axios2.test.js:
```diff
@@ -1271,4 +1271,182 @@ describe("Axios AuthHttpRequest class tests", function() {
+        }
+    });
+
+    it("test when ACCESS_TOKEN_PAYLOAD_UPDATED is fired", async function() {
```
```suggestion
    it("test ACCESS_TOKEN_PAYLOAD_UPDATED when updated with handle", async function() {
```
